### ISSUE TITLE
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted

### ISSUE THREAD

**Issue by j-ed** [nextcloud/previewgenerator#108]
## Expected behavior
The preview generation task shouldn't use all PHP memory,

## Current behavior
If the preview generation is run, by e.g. executing the command `./occ preview:generate-all <user>`, the following PHP error might be shown. From my point of view it need to make sure that unused memory is freed instantly after the preview images have been created. Increasing the available PHP memory, as recommended here https://github.com/rullzer/previewgenerator/pull/106, is only half of the job and shouldn't replace a thrifty memory handling.
```
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 9792 bytes) in /srv/www/nextcloud/nextcloud/lib/private/legacy/image.php on line 496
```

### Steps to reproduce
1. Open a command line window
2. Execute: /occ preview:generate-all <user>

#### Server Configuration
OS: Linux 3.16.54
Web server: Apache2 2.4.29
Database: MariaDB 10.0.32
PHP version: 5.6.29
Nextcloud version: 12.0.5

#### Client Configuration
Browser: Mozilla Firefox 58.0.2
Operating system: Windows 10

**Comment by HanZ92** [comment]
It's quite problematic, since of my users has a image file, which is around 34MB in size. I am not able to let Preview Generator finish it's job, because it doesn't matter, which value i put my memory_limit on, it still aborts. Even tried it with -1, it still aborts. It needs to make sure, that when it gets an error like this, it continues with the next file.

**Comment by arno01** [comment]
This is likely because you have data segment size limit (ulimit -d) is set low (134217728 bytes), try setting ``ulimit -d unlimited`` for the account you are starting the Nextcloud with. Or you can set it in the Nextcloud's systemd service like this:

```
[Service]
LimitDATA=unlimited
```

**Comment by sirrahd** [comment]
I run it like this, which I think would satisfy your needs, increase the limit for thumbnail generation only:
`php -d memory_limit=1024M occ preview:generate-all`

**Comment by j-ed** [comment]
@arno01 That was one of the first things I checked, unfortunately the parameter is correctly set but the problem remains: 
```
 > ulimit -d
unlimited
```

**Comment by meyerf99** [comment]
@j-ed Did you fixed the problem? 
I have the same problem on my Synologie RS812+ with round about 1.8k files.

**Comment by HanZ92** [comment]
@meyerf99 How do you call the preview generator?

**Comment by HanZ92** [comment]
I run it with this command for PHP 7.0

` /bin/su http -s /bin/sh -c '/usr/local/bin/php70 -c /usr/syno/etc/packages/WebStation/php70/conf.d/user_settings.ini -f /volume1/web/nextcloud/occ preview:pre-generate'`

And then I set under `/usr/syno/etc/packages/WebStation/php70/conf.d/user_settings.ini` the value accordingly.

But nevertheless: You normally need to call the command from the http user, not the current user.

My PHP 5.6 command for the cron is therefore like this:
`/bin/su -s /bin/sh -c "/usr/local/bin/php56 -f /volume1/web/nextcloud/cron.php" http`

You can check if it works with this syntax.

But the first PHP 7.0 code runs fine for me with over 500.000 files.

**Comment by j-ed** [comment]
@meyerf99 The problem hasn't been reall y solved yet. I first reduced the number of previews to be created, as recommended here: https://github.com/rullzer/previewgenerator . This reduced the amont of disk space dramatically and speeded-up the preview loading in the mobile app. Then I've called the command to generate previews multiple times until all previews had successfully been created.

**Comment by meyerf99** [comment]
@HanZ92 I tried with your PHP 7.0 command but it doesn't work
Now try with PHP 5.6 and Memory Limit up to 1028M.

**Comment by mcnesium** [comment]
What is the proper amount of memory to set for this? I gave it 4G and still it complains about exausted memory size:

    Cron <root@srv>  docker exec -u www-data -i nextcloud php -d memory_limit=4096M occ preview:pre-generate

    PHP Fatal error:  Allowed memory size of 4294967296 bytes exhausted (tried to allocate 23622984 bytes) in /var/www/html/lib/private/legacy/OC_Image.php on line 649

**Comment by rusitschka** [comment]
There seems to be a memory leak in preview:generate-all. It allocates 100-200 MB per second and doesn't free those. Here's a /proc/meminfo trace which prints MemFree every second:

```
$ while true; do grep MemFree /proc/meminfo; sleep 1; done
MemFree:         3272776 kB
MemFree:         3173776 kB
MemFree:         3072056 kB
MemFree:         2955836 kB
MemFree:         2794884 kB
MemFree:         2714732 kB
MemFree:         2594080 kB
MemFree:         2473300 kB
MemFree:         2355804 kB
MemFree:         2235112 kB
MemFree:         2113032 kB
MemFree:         1983352 kB
MemFree:         1909972 kB
MemFree:         1766320 kB
MemFree:         1715428 kB
MemFree:         1587836 kB
MemFree:         3690872 kB
```

During that run 10-20 images are processed. Then the script OOMs. Tried 2 GB memory limit and 4 GB. Same thing...

**Comment by joshuata** [comment]
I see the same issue. Some more details:

* The memory usage doesn't seem to rise while iterating files. This likely rules out a leaky directory iterator
* There are occasional stair-step decreases in allocated memory, likely corresponding to garbage collection
* The memory leak is not always there. It doesn't occur on every directory, and if it recovers once (with a GC round), its unlikely to freeze again

My guess is that the leak is in the returned [IPreview object in the parseFile function](https://github.com/rullzer/previewgenerator/blob/0b2cf052afd609f422d62d479aaf80e51fd2c1df/lib/Command/Generate.php#L184). This is my best guess since these objects are not allocated if the preview already exists, but must be allocated if not. That fits the observed pattern fairly well, and dangling file handles or open stream objects are frequent sources of this form of memory leak.

**Comment by joshuata** [comment]
I am now almost certain that the issue is [this one](https://github.com/nextcloud/server/issues/19022) in nextcloud server. The filesize cutoff matches exactly what I'm seeing on my server.

**Comment by elijahnyp** [comment]
I tracked down where I was having this problem - OOM on generating MANY preview images.  Turns out, for me, it was a simple fix in the nextcloud preview generating code.  Don't know if this is helpful, or even applicable here.

https://github.com/nextcloud/server/issues/21995

There may be a performance hit in previewgenerator but this may be fixable by having previewgenerator call generatePreviews directly with multiple specifications instead of calling getPreview for each configured size.

**Comment by rclough** [comment]
using `-d 1024M` (or whatever other memory value) option safe or will it prevent generation of certain images? because it seems to work for me

**Comment by joshuata** [comment]
It does not work if there are lots of thumbnails to generate. I've been running it with about 8 GB of memory, and I still run out. I have thousands of large-ish photos, so it's worst case, but since the memory is almost never garbage collected it will reliably hit OOM after 30 or so images

**Comment by MorrisJobke** [comment]
Closed via nextcloud/server#22008

**Comment by joshtrichards** [comment]
This Issue needs to be closed out but I lack permissions in this repository to do so. :)
